src/DependencyInjection/DI/perf/ConstrainedGenericBenchmark.cs

```diff
@@ -0,0 +1,346 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using BenchmarkDotNet.Attributes;
+
+namespace Microsoft.Extensions.DependencyInjection.Performance
+{
+    public class ConstrainedGenericBenchmark
```

Comment:
This not exactly what I'd like to be benchmarked. In this benchmark, we are exercising resolution path but all the changes are in the call site creation code path.

What I would like to see is a benchmark that uses CallSiteFactory directly and measures how long does it take to build a call site with a generic constraint before and after this change.

Reply:
So basically like the `CallSiteFactoryTest`'s? I can do that. What's the before/after look like to you? Literally executing the same benchmarks on before/after all my changes? All of the one with constraints will just throw exceptions, is that what you're looking for?

Reply:
I'll change the NoDI one to be NoConstraints, too.

Reply:
And there's just one branch that goes into the constraints stuff, so I can just remove that branch and run the benchmarks for the "before" picture.

Reply:
> So basically like the CallSiteFactoryTest

Yes

There are two goals here:

1. Make sure we didn't regress perf for existing scenarios
2. Make sure we have reasonable perf for new ones

Running benchmarks with constraints that are satisfied before and after the change would show us if it made things slower in existing cases and we would have something to compare the performance of new scenarios too.

Reply:
Here's an example benchmark:

```c#
[GlobalSetup(Target = nameof(NoConstraints))]
public void SetupNoConstraints()
{
    _noConstraintsCsf = new CallSiteFactory(new[]
    {
        new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(A<>)),
        new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(B<>)),
        new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(C<>)),
        new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(D<>)),
        new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(E<>)),
        new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(F<>)),
        new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(G<>)),
        new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(H<>)),
    });
}

[Benchmark(OperationsPerInvoke = OperationsPerInvoke)]
public void NoConstraints()
{
    for (int i = 0; i < OperationsPerInvoke; i++)
    {
        var _ = _noConstraintsCsf.GetCallSite(typeof(IEnumerable<IFakeGeneric<M>>), new CallSiteChain());
    }
}
```

This about right?

Reply:
You can reuse the same `_noConstraintsCsf` field. And I think you don't need multiple implementation types for this case, registering the same type multiple times would do. Iteration time for this benchmark would be fairly substantial so you can drop `OperationsPerInvoke`

Another thing to consider is that `GetCallSite` caches result in a dictionary so only first iteration would measure what we want while all others would just measure dictionary retrieval. You would have to move `new CallSiteFactory` into benchmark itself to run code in question every iteration.

Reply:
Something like this then:

```c#
[Benchmark(OperationsPerInvoke = OperationsPerInvoke)]
public void NoConstraints()
{
    for (int i = 0; i < OperationsPerInvoke; i++)
    {
        var csf = new CallSiteFactory(new[]
        {
            new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(NoConstraintsOpenGenericService<>)),
            new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(NoConstraintsOpenGenericService<>)),
            new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(NoConstraintsOpenGenericService<>)),
            new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(NoConstraintsOpenGenericService<>)),
            new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(NoConstraintsOpenGenericService<>)),
            new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(NoConstraintsOpenGenericService<>)),
            new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(NoConstraintsOpenGenericService<>)),
            new ServiceDescriptor(typeof(IFakeGeneric<>), typeof(NoConstraintsOpenGenericService<>)),
        });
        var _ = csf.GetCallSite(typeof(IEnumerable<IFakeGeneric<M>>), new CallSiteChain());
    }
}
```

Reply:
Remove OperationsPerInvoke and yeah let's put ServiceDescriptor array into a static

Reply:
Before:

https://gist.github.com/jbogard/daa09b6a9741e34806986b516fc6f147

After:

https://gist.github.com/jbogard/355f85195a4008bdf61f61284d579325

Reply:
Fixed the benchmarks to have static fields